### ISSUE TITLE
pyright does not raise the same error as mypy to `Tuple[float, ...]` (type hint with ellipsis)

### ISSUE THREAD

**Issue by ftnext** [microsoft/pyright#1172]
**Describe the bug**

Using type hinting with ellipsis `Tuple[float, ...]`, mypy shows an error, whereas pyright does not show any error.

It seems like a bug to me.
But I may have misunderstood something because I'm just starting to learn the type hinting.
Thanks!

**To Reproduce**

```
(env38) $ sw_vers
ProductName:	Mac OS X
ProductVersion:	10.14.6
BuildVersion:	18G4032
(env38) $ python -V
Python 3.8.6
(env38) $ node --version
v12.13.0
(env38) $ npm --version
6.12.0
(env38) $ ./node_modules/.bin/pyright --version
pyright 1.1.85

(env38) $ ./node_modules/.bin/pyright pyright_elipsis.py -v env38
No configuration file found.
stubPath /Users/.../typings is not a valid directory.
Assuming Python platform Darwin
Searching for source files
Found 1 source file
0 errors, 0 warnings, 0 infos
Completed in 0.548sec
```

**Expected behavior**
Raise error like mypy

```
(env38) $ mypy --version
mypy 0.790
(env38) $ mypy pyright_elipsis.py
pyright_elipsis.py:7: error: Incompatible types in assignment (expression has type "Tuple[float, float, float, str]", variable has type "Tuple[float, ...]")
Found 1 error in 1 file (checked 1 source file)
```

**Screenshots or Code**

pyright_elipsis.py

```python
from typing import Tuple

Vector = Tuple[float, ...]

vector: Vector = (3.0, 9.8)
vector = (1.0, -4.2, 5.4)
vector += ("hoge",)
```

**VS Code extension or command-line**

a command-line tool.
I wrote version info at *To reproduce* section.

**Additional context**

References
https://docs.python.org/3.8/library/typing.html#typing.Tuple
>To specify a variable-length tuple of homogeneous type, use literal ellipsis, e.g. Tuple[int, ...].

**Comment by erictraut** [comment]
Thanks for the bug report.

Based on the type information within the typeshed stdlib stub, pyright is doing the right thing. Here's the declaration for `tuple` (with portions omitted for brevity):

```python
class tuple(Sequence[_T_co], Generic[_T_co]):
    ...
    @overload
    def __add__(self, x: Tuple[_T_co, ...]) -> Tuple[_T_co, ...]: ...
    @overload
    def __add__(self, x: Tuple[Any, ...]) -> Tuple[Any, ...]: ...
```

It supports two overloads for the `__add__` magic method. The first applies in cases where the type argument matches. For example, in the expression `vector + (3.4, )`. The second overload applies in all other cases, and it produces a value with the resulting type of `Tuple[Any, ...]`. The `Any` type matches any type, hence a `Tuple[Any, ...]` is assignable to a variable of type `Tuple[float, ...]`.

I suspect what's happening here is that someone has added special-case knowledge of the `tuple` class within mypy, and this logic is overriding the type information provided in the stub file. In general, that's something we don't like to do in pyright. But `tuple` is an exceptional case because it's such a fundamental type in Python, and it's one of the only types that supports variadic (i.e. a variable number of) type arguments. 

I'll think more about whether we want to make an exception and add special-case logic specifically for the `__add__` method of the `tuple` class.

**Comment by erictraut** [comment]
I've looked at this more. It would require a couple of pretty hacky special-case blocks of code within the type checker. It's a slippery slope to add special-case logic that overrides the knowledge within the stdlib type stub files. If this were a false positive, I could maybe justify adding it. But given that it's a false negative — and one that probably doesn't show up very frequently, I don't think the extra special-case logic is worth it.
